ujt/callbacks/change_over_time_callbacks.py

```diff
@@ -38,20 +50,171 @@ def update_time_select_panel(tag):
+        tag_selection: The value from the CHANGE_OVER_TIME_TAG_DROPDOWN. Should be CUSTOM_TIME_RANGE_DROPDOWN_VALUE or a timestamped tag.
+        sli_type; The value from the CHANGE_OVER_TIME_SLI_TYPE_DROPDOWN. Used for input validation.
+
+    Returns:
+        A dictionary to be placed in the CHANGE_OVER_TIME_SLI_STORE.
+        The dictionary has keys "start_timestamp" and "end_timestamp", which map
+        to the POSIX timestamps of their respective times.
+    """
+
+    ctx = dash.callback_context
+    triggered_id, triggered_prop, triggered_value = utils.ctx_triggered_info(ctx)
+
+    if triggered_id == id_constants.CHANGE_OVER_TIME_RESET_BUTTON:
+        return [{}, dash.no_update, dash.no_update]
+
+    if sli_type is None:
+        return [dash.no_update, True, "Please select a SLI type."]
+
+    try:
+        if tag_selection == constants.CUSTOM_TIME_RANGE_DROPDOWN_VALUE:
+            start_time = dt.datetime.strptime(
+                start_time_input_values[0], constants.DATE_FORMAT_STRING
+            )
+            end_time = dt.datetime.strptime(
+                end_time_input_values[0], constants.DATE_FORMAT_STRING
+            )
+        else:
+            temp_datetime = dt.datetime.strptime(
+                window_size_input_values[0], constants.WINDOW_SIZE_FORMAT_STRING
+            )
+            window_size_timedelta = dt.timedelta(
+                hours=temp_datetime.hour,
+                minutes=temp_datetime.minute,
+                seconds=temp_datetime.second,
+            )
+
+            tag_timestamp_string = tag_selection.split("@")[-1]
+            tag_timestamp = dt.datetime.strptime(
+                tag_timestamp_string, constants.DATE_FORMAT_STRING
+            )
+            start_time = tag_timestamp - window_size_timedelta
+            end_time = tag_timestamp + window_size_timedelta
+    except (ValueError, TypeError):
+        # ValueError occurs when input format is incorrect
+        # TypeError occurs when input box is blank
+        return dash.no_update, True, "Error parsing time input, please check format."
+
+    sli_response = rpc_client.get_slis(
+        start_time=start_time,
+        end_time=end_time,
+        sli_types=[sli_type],
+    )
+
+    dict_slis: List[Dict[str, Any]] = [
+        json_format.MessageToDict(sli) for sli in sli_response.slis
+    ]
+
+    # Dash serializes the objects placed into Stores via json, but this can lead to
+    # unexpected behavior with objects (particularly datetimes).
+    # This can make it confusing to store objects directly, so we choose to store
+    # timestamps instead of datetimes for a clearer interface from the reader's perspective.
+    # We are forced to convert SLI protos to dictionaries to store them as well.
+    return [
+        {
+            "start_timestamp": start_time.timestamp(),
+            "end_timestamp": end_time.timestamp(),
+            "dict_slis": dict_slis,
+        },
+        False,
+        None,
+    ]
```

Comment:
We let this callback (which is fired when the query or clear button is clicked) make the RPC to get SLIs. We then save the SLI results to a component.

This is necessary to enable multiple components to get the SLIs without making repeated network calls (whose results may not be consistent even if we provide the same arguments). In this case, I want the graph's style to change based on the SLIs, and a datatable to display the SLIs. Thus, the only way to share this data is through saving it to this component.

We experience a noticeable performance penalty when we serialize this list of SLIs, which becomes worse as the time range increases. However, I don't see any viable alternatives to this approach, and hopefully the "query" language in the button makes it obvious to users that there could be a delay.